### ISSUE TITLE
Should the custom loss function in Keras return a single loss value for the batch or an arrary of losses for every sample in the training batch?

### ISSUE THREAD

**Issue by lambdaphy** [tensorflow/tensorflow#42446]
I asked a question on [StackOverflow](https://stackoverflow.com/questions/63390725/should-the-custom-loss-function-in-keras-return-a-single-loss-value-for-the-batc) regarding as the return value of a custom loss funtion. But I didn't get a clear answer.

In this [guide](https://www.tensorflow.org/guide/keras/train_and_evaluate#custom_losses) on tensorflow website, I found an example of custom loss funciton:

        def custom_mean_squared_error(y_true, y_pred):
            return tf.math.reduce_mean(tf.square(y_true - y_pred))

The `reduce_mean` function in this custom loss function will return an scalar. But I think the custom loss function should return an array of losses for every example in a training batch, rather than a single loss value. 

According to the source code of [Model](https://github.com/tensorflow/tensorflow/blob/v2.3.0/tensorflow/python/keras/engine/training.py#L159-L2634) class, the custom loss function is used to constructed a `LossFunctionWrapper` object. I read the source code of the [loss](https://github.com/tensorflow/tensorflow/blob/v2.3.0/tensorflow/python/keras/losses.py) module. I think it's `LossFunctionWrapper.__call()__` method that is responsible for getting the mean loss value for the training batch. `LossFunctionWrapper.__call()__` method first calls the `LossFunctionWrapper.call()` method to get an array of losses for every example in the training batch. It's in the  `LossFunctionWrapper.call()` method that our custom loss function is called.

In addition, in the souece code of [losses](https://github.com/tensorflow/tensorflow/blob/v2.3.0/tensorflow/python/keras/losses.py) module, the `MeanAbsoluteError` class uses the `mean_squared_error` function to construct a `LossFunctionWrapper` class. We can see that the `mean_squared_error` function returns `K.mean(math_ops.squared_difference(y_pred, y_true), axis=-1)`, which is an array, not a single value. I think our custom loss function shoud just be like this.

So, why does the custom loss function in the [guide](https://www.tensorflow.org/guide/keras/train_and_evaluate#custom_losses) on tensorflow website return a scalar? Is it wrong to define a custom function like this?

**Comment by jvishnuvardhan** [comment]
@lambdaphy custom loss function in keras returns a single value as in the above example we are using `tf.math.reduce_mean`. Simple example of reduce_mean is as follows.

```
import tensorflow as tf
a = tf.ones([1,100],tf.int32)
reduce_m = tf.math.reduce_mean(a) 
print(reduce_m)       # output tf.Tensor(1, shape=(), dtype=int32)

```
`reduce_mean` of the above array `a` is 1 as shown in the output. 

For Keras models, we need to define those custom_loss functions to provide a scalar loss for each batch. For custom_training, you can define loss functions as you want to implement in your custom model. Thanks!

**Comment by lambdaphy** [comment]
I know that when training a model we need a scalar loss for each batch. But according to the source code, the custom loss function is not responsible for getting this scalar loss. Could you check the source codes of the `Model.compile()`, `Model.fit()` , `Loss.__call()__` , `Loss.call()` and  `LossFunctionWrapper.call()` methods to look the calculating process of loss value of a training batch? 

I read those codes, and I find the `Loss.__call__()` method calls the `Loss.call()` method (which is implemented in subclass, e.g. `LossFunctionWrapper`) to get an array of losses for every example in the training batch, then `Loss.__call__()` method calls `compute_weighted_loss()` function to get the (weighted) average of those losses.  How does the `LossFunctionWrapper.call()` method get an array of losses? Reading the source code, we can see that it uses `self.fn` to get those losses, and the `self.fn` is the loss funciton we provided to `Model.compile()` method.

Here is the source code of `Loss.__call__()`:

        def __call__(self, y_true, y_pred, sample_weight=None):
           """
           .................(omitted)
           Returns:
                Weighted loss float `Tensor`. If `reduction` is `NONE`, this has
                shape `[batch_size, d0, .. dN-1]`; otherwise, it is scalar. (Note `dN-1`
                because all loss functions reduce by 1 dimension, usually axis=-1.)
           .................(omitted)
           """
            graph_ctx = tf_utils.graph_context_for_symbolic_tensors(
                    y_true, y_pred, sample_weight)
            with K.name_scope(self._name_scope), graph_ctx:
            ag_call = autograph.tf_convert(self.call, ag_ctx.control_status_ctx())
            losses = ag_call(y_true, y_pred)
            return losses_utils.compute_weighted_loss(
                          losses, sample_weight, reduction=self._get_reduction())

Here is the source code of `Loss.call()`:

        def call(self, y_true, y_pred):
        """Invokes the `Loss` instance.
            Args:
                y_true: Ground truth values. shape = `[batch_size, d0, .. dN]`, except
                sparse loss functions such as sparse categorical crossentropy where
                shape = `[batch_size, d0, .. dN-1]`
                y_pred: The predicted values. shape = `[batch_size, d0, .. dN]`
           Returns:
               Loss values with the shape `[batch_size, d0, .. dN-1]`.
        """
            NotImplementedError('Must be implemented in subclasses.')

The `Loss.call()` method is just an interface that a subclass of `Loss` must implement. But we can see that the return value of this method is **Loss values with the shape `[batch_size, d0, .. dN-1]`**.

Now let's see `LossFunctionWrapper` class. `LossFunctionWrapper` is a subclass of `Loss`. In its constructor, we should provide a loss function, which is stored in `LossFunctionWrapper.fn`. Here is the source code of `LossFunctionWrapper.call()`, which implements `Loss.call()` method:

        def call(self, y_true, y_pred):
        """Invokes the `LossFunctionWrapper` instance.
            Args:
                 y_true: Ground truth values.
                 y_pred: The predicted values.
            Returns:
                Loss values per sample.
        """
             if tensor_util.is_tensor(y_pred) and tensor_util.is_tensor(y_true):
                 y_pred, y_true = tf_losses_util.squeeze_or_expand_dimensions(y_pred, y_true)
             ag_fn = autograph.tf_convert(self.fn, ag_ctx.control_status_ctx())
             return ag_fn(y_true, y_pred, **self._fn_kwargs)

See? The loss function we provide is called here, and its task is to return **Loss values per sample**.

In addition, as an example, we can see how the `MeanSquaredError` class is defined. It is just a `LossFunctionWrapper` class using the `mean_squared_error` function as the loss function(i.e. `LossFunctionWrapper.fn=mean_squared_error`). The source code of `mean_squared_error` function is also defined in the `losses` module:

    def mean_squared_error(y_true, y_pred):
        """Computes the mean squared error between labels and predictions.
            After computing the squared distance between the inputs, the mean value over
            the last dimension is returned.
           `loss = mean(square(y_true - y_pred), axis=-1)`
            Standalone usage:
            >>> y_true = np.random.randint(0, 2, size=(2, 3))
            >>> y_pred = np.random.random(size=(2, 3))
            >>> loss = tf.keras.losses.mean_squared_error(y_true, y_pred)
            >>> assert loss.shape == (2,)
            >>> assert np.array_equal(
                     ...     loss.numpy(), np.mean(np.square(y_true - y_pred), axis=-1))
           Args:
              y_true: Ground truth values. shape = `[batch_size, d0, .. dN]`.
              y_pred: The predicted values. shape = `[batch_size, d0, .. dN]`.
           Returns:
             Mean squared error values. shape = `[batch_size, d0, .. dN-1]`.
        """
        y_pred = ops.convert_to_tensor_v2(y_pred)
        y_true = math_ops.cast(y_true, y_pred.dtype)
        return K.mean(math_ops.squared_difference(y_pred, y_true), axis=-1)

We can see that it returns an array, not a scalar (the `axis-1` in the `K.mean()`), and the firse dimension of its return value is `batch_size`. 

According the the source code of `Model.compile()` and `Model.fit()`, when we provide a custom loss function, this funtion is used to construct a `LossFunctionWrapper` object, just like the `MeanSquaredError` object uses `mean_squared_error` fucntion to construct a `LossFunctionWrapper` object. That's why I think that the custom loss function shoud return an array of losses, because obtaining a scalar loss value for the training batch is not the task of a loss funtion, the `Loss.__call()__` shoud do that job.

**Comment by lambdaphy** [comment]
Based on the above analysis, we can see that when we define our custom loss **CLASS**, we should implement the `call()` method, and this `call()` method SHOULD return an array, not a scalar. But in the [guide](https://www.tensorflow.org/guide/keras/train_and_evaluate#custom_losses) on tensorflow, we can also see an example of custom loss class, just below the example of custom loss function:

    class CustomMSE(keras.losses.Loss):
        def __init__(self, regularization_factor=0.1, name="custom_mse"):
            super().__init__(name=name)
            self.regularization_factor = regularization_factor

        def call(self, y_true, y_pred):
            mse = tf.math.reduce_mean(tf.square(y_true - y_pred))
            reg = tf.math.reduce_mean(tf.square(0.5 - y_pred))
            return mse + reg * self.regularization_factor

    model = get_uncompiled_model()
    model.compile(optimizer=keras.optimizers.Adam(), loss=CustomMSE())

    y_train_one_hot = tf.one_hot(y_train, depth=10)
    model.fit(x_train, y_train_one_hot, batch_size=64, epochs=1)

We can see that this implementation of `call()` method returns a scallar. I think this behavior is also wrong.

**Comment by pavithrasv** [comment]
@lambdaphy Thank you for the issue. Custom loss function is required to return one loss value per sample. The example will need to be updated to reflect this.

**Comment by pavithrasv** [comment]
If you are interested in making the change, please feel free to send me a PR.

**Comment by lambdaphy** [comment]
Thank you for your reply. I'm very happy to contribute, but I am not familiar with using github to open pull request. In addition I'm not a native English speaker, so maybe I'm not competent to edit the documentation.

If you can edit the documentation, just help change it. Thanks.

**Comment by lambdaphy** [comment]
> Most of guidances follow the tf doc which return single value for a batch of training data. I follow the source code to the `compute_weighted_loss` where `LossFunctionWrapper` do the mean across batch axis, and found that if custom loss func return a value instead of a array of value, we got two problems:
> 
> 1. `sample_weights` argument will be meaningless since wo do mean first.
> 2. Loss are averaged based on a uncorrect denominator when do training with `tf.distribute.Strategy`.
>    @lambdaphy what do you think?

Yes, you're right.  Loss function returning a single loss value for the whole batch will cause problems under some circumstances.

The documentation has yet to be corrected.

There are many ambiguities in the document.

**Comment by AshwinJay101** [comment]
Okay the example has still not been updated

The conclusion is that it should return an array rather than a single value?

**Comment by TalhaUsf** [comment]
> Okay the example has still not been updated
> 
> The conclusion is that it should return an array rather than a single value?

**It should return array of sample losses.** 

So following is wrong : 🚫 
```
def custom_mean_squared_error(y_true, y_pred):
        return tf.math.reduce_mean(tf.square(y_true - y_pred))
```

Following is correct: ✅ 

```
def custom_mean_squared_error(y_true, y_pred):
        return tf.math.reduce_mean(tf.square(y_true - y_pred), axis = -1)
```

**Comment by AshwinJay101** [comment]
Thanks @lambdaphy 

We should update this [guide](https://www.tensorflow.org/guide/keras/train_and_evaluate#custom_losses)

I can update it if we all agree with that

**Comment by mromanelloj** [comment]
I am still confused. If we are trying to produce an array with one loss value for each sample within a batch, wouldn't we use `def custom_mean_squared_error(y_true, y_pred):
    return tf.math.reduce_mean(tf.square(y_true - y_pred), axis = 1))`
since the batch size is the first dimension?

**Comment by tilakrayal** [comment]
@lambdaphy,
The related PR #https://github.com/keras-team/keras-io/pull/1288 was merged and also the requested changes were also reflected on the tensorflow.org and Keras.io official documents.
https://www.tensorflow.org/guide/keras/training_with_built_in_methods#custom_losses
https://keras.io/guides/training_with_built_in_methods/



**Before:**

```
def custom_mean_squared_error(y_true, y_pred):
        return tf.math.reduce_mean(tf.square(y_true - y_pred))
```

**After:**
```
def custom_mean_squared_error(y_true, y_pred):
        return tf.math.reduce_mean(tf.square(y_true - y_pred), axis = -1)
```

Could you please check the source file https://github.com/keras-team/keras-io/blob/master/guides/ipynb/training_with_built_in_methods.ipynb and confirm the same. Thank you!
